### PR TITLE
Change minimum version of Python from 3.8 to 3.9

### LINKED ISSUE #1458: Intermittent test failure in `TestSyntheticRadiograph.test_optical_density_histogram`
In #1456, I got a test failure for the Python 3.8 (Windows) check for `TestSyntheticRadiograph.test_optical_density_histogram`.  The test failure only showed up in one check, and did not show up in any of the other checks.  

I'm wondering if specifying a random seed for this test would enable us to resolve the problem, and also make the test deterministic.

Here's the test failure message:

```
==================================================================================== FAILURES =====================================================================================
_____________________________________________________________ TestSyntheticRadiograph.test_optical_density_histogram ______________________________________________________________
[gw0] win32 -- Python 3.8.10 D:\a\PlasmaPy\PlasmaPy\.tox\py38-all\Scripts\python.EXE

self = <plasmapy.diagnostics.tests.test_charged_particle_radiography.TestSyntheticRadiograph object at 0x0000023CE26F7AF0>

    def test_optical_density_histogram(self):
        """
        Test the optical density calculation is correct and stuffed
        with numpy.inf when the intensity is zero.
        """
        bins = (200, 60)
        size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
    
        intensity_results = cpr.synthetic_radiograph(
            self.sim_results, size=size, bins=bins
        )
        od_results = cpr.synthetic_radiograph(
            self.sim_results, size=size, bins=bins, optical_density=True
        )
    
        assert np.allclose(intensity_results[0], od_results[0])
        assert np.allclose(intensity_results[1], od_results[1])
    
        intensity = intensity_results[2]
        zero_mask = intensity == 0
        i0 = np.mean(intensity[~zero_mask])
        od = -np.log10(intensity / i0)
    
>       assert np.allclose(od[~zero_mask], od_results[2][~zero_mask])
E       assert False
E        +  where False = <function allclose at 0x0000023CD57F8B80>(array([ 5.88380294e-01,  6.50528201e-01,  3.66531544e-01,  8.59669742e-03,\n       -9.88513070e-02,  4.42252258e-01,  1...7155e+00,  4.22048872e-01, -1.27623050e-01,\n       -8.67867956e-02,  3.66531544e-01,  7.64471553e-01,  8.10229044e-01]), array([ 5.85862638e-01,  6.48010545e-01,  3.64013888e-01,  6.07904120e-03,\n       -1.01368963e-01,  4.39734602e-01,  1...5390e+00,  4.19531216e-01, -1.30140706e-01,\n       -8.93044518e-02,  3.64013888e-01,  7.61953897e-01,  8.07711387e-01]))
E        +    where <function allclose at 0x0000023CD57F8B80> = np.allclose

plasmapy\diagnostics\tests\test_charged_particle_radiography.py:598: AssertionError
```

### LINKED ISSUE #1796: Support Python 3.9+ for releases after April 2023
[NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) proposes a drop schedule for old versions of NumPy and Python.  NEP 29 calls for support of Python 3.8 through releases made up until April 14, 2023.  Since NEP 29 has emerged as a community standard, it makes sense for us to follow it as well.  In particular, packages that we strong depend on (NumPy, Astropy, etc.) are following that schedule.

For releases after mid-April 2023, we should require Python ≥3.9 and drop Python 3.8 support.  This will probably be for PlasmaPy ≥0.11 if we want to make releases every ∼3–4 months.  

When we do this, we would be able to automagically apply a lot of the improved syntax by changing `--py38-plus` to `--py39-plus` in `.pre-commit-config.yaml`.  We should consider dropping `--keep-runtime-typing` as well for postponed evaluation of annotations.  I did a test run, and most of the changes from `pyupgrade` were things like `typing.List` → `list` in type hint annotations.

We'd also be able to start using [`typing.Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated), which could potentially help with decorators like `@particle_input` and let us use physical types in type hints as well.

Here's the link to [what's new in Python 3.9](https://docs.python.org/3/whatsnew/3.9.html).

This is the long awaited sequel to #1262.

### PR DESCRIPTION
**Pull request by namurphy**
## Description

This PR changes package, testing, and pre-commit config files and applies changes from `pyupgrade` and `ruff` for the switch from Python 3.8+ to Python 3.9+. 

The changes from `pyupgrade` and `ruff` are primarily related to type hint annotations (e.g., changing `typing.List[str]` → `list[str]`).

This PR should be reviewed and merged after the release of `v2023.1.0`.

## Motivation and context

[NumPy Enhancement Proposal 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) lays out a recommended community schedule for dropping support of old versions of packages.  As per the schedule, Python 3.8 support can be dropped for releases after April 14, 2023.

## Related issues

Closes #1796. See also #912, #1661, #1883

Cross-references: #1262, #980, #1320

Closes #1458, hopefully.  #1458 was about the seemingly Python 3.8 only intermittent test failure in `TestSyntheticRadiograph`.

### CODE REVIEW

**Comment by namurphy** [comment]
- [ ] Change the GitHub repo setting to not require 'Python 3.8, all tests, code coverage'

File: pyproject.toml [hunk 1]
```diff
@@ -17,26 +17,26 @@ keywords = [
...
   "ipywidgets>=7.6.5",
   "lmfit>=1",
   "matplotlib>=3.3",
   "mpmath>=1.2.1",
   "numba",
-  "numpy>=1.20",
+  "numpy>=1.21",
   "packaging",
-  "pandas>=1",
+  "pandas>=1.2",
```
Comment (namurphy) [inline pyproject.toml:32]: I think it was the `pandas` dependency that was causing problems on the tests with the minimum versions.  For most of the dependencies that I changed, I looked for a release that was roughly two years old and set that as the minimum version, which is what NEP 29 recommends doing for NumPy and is probably a good balance between reducing maintenance burden and having a wide enough window of allowed package versions.

**Comment by namurphy** [comment]
**Important note:** I temporarily turned off the branch protection rule for requiring the Python 3.8 minimal dependencies check to pass before merging.  After this PR is merged, we'll need to re-enable it on the Python 3.9 equivalent (and perhaps some of the other checks too).  We'll need to do the same thing for switching the tests to nox as well.

File: docs/common_links.rst [hunk 2]
```diff
@@ -104,7 +104,7 @@
 .. |contributor guide| replace:: :ref:`contributor guide`\
 .. |documentation guide| replace:: :ref:`documentation guide`\
 .. |glossary| replace:: :ref:`glossary`\
-.. |minpython| replace:: 3.8
+.. |minpython| replace:: 3.9
```
Comment (StanczakDominik) [inline docs/common_links.rst:107]: Wow.

File: plasmapy/analysis/fit_functions.py [hunk 3]
```diff
@@ -32,21 +32,21 @@ class AbstractFitFunction(ABC):
...
 
-    _param_names = NotImplemented  # type: Tuple[str, ...]
+    _param_names = NotImplemented  # type: tuple[str, ...]
 
     def __init__(
         self,
-        params: Tuple[float, ...] = None,
-        param_errors: Tuple[float, ...] = None,
+        params: tuple[float, ...] = None,
+        param_errors: tuple[float, ...] = None,
```
Comment (StanczakDominik) [inline plasmapy/analysis/fit_functions.py:40]: Shouldn't we also go through `future` imports, btw? see if we can't remove them? Or would pyupgrade remove them by itself?

**Review by StanczakDominik — CHANGES_REQUESTED** [review]
LGTM! A couple minor issues. I'll run through this to check the `future` imports.

File: plasmapy/analysis/fit_functions.py (hunk 3 above)
Comment (StanczakDominik) [reply plasmapy/analysis/fit_functions.py:40]: Ah. Guess not; the only `__future__` imports we have are `from __future__ import annotations`, and [that still isn't going to be the default for a while](https://docs.python.org/3/library/__future__.html#id2).

**Review by StanczakDominik — APPROVED** [review]
Applied the changelog merge; feel free to revert it if you prefer doing them separately, though.

File: changelog/1885.removal.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Changed the minimum required version of Python from 3.8 to 3.9.
```
[the remark this answers is not in the document]
Comment (namurphy) [reply changelog/1885.removal.rst]: Yeah, I wasn't sure where to put this one!  Happy to go with `breaking` instead.

**Comment by namurphy** [comment]
Thank you for the reviews!

**Pull request merged**